### ISSUE TITLE
USWDS should provide React components

### ISSUE THREAD

**Issue by thisisdano** [uswds/uswds#2884]
## Description

Include a high-level description of the issue. What did you expect to happen? What happened instead? What would you like to see changed?

Include any benefits, challenges, or considerations. This can be short and sweet.

## Steps to reproduce the issue

1. Step one
2. Step two
3. Step three
4. etc.

## Additional information [optional]

* Relevant research and support documents
* Screen shot images
* Notes
* And so on

**Comment by davemcorwin** [comment]
The goal should be to be able to use the USWDS with a variety of frameworks without too much work, but not to tie it to a specific one such as React. That said, we can use React as a test bed for understanding what changes need to be made and where to draw the lines.

The problems listed in the issue description are correct, I would add that the components' JS are not isolated, many of them reach out to the `window`, `document`, or `body` AND because we want to make sure these remain accessible is may be hard to isolate the correct pieces.

There is not a whole lot of JS here to begin with, so it might be possible to re-implement in various frameworks, but the maintenance still seems like it would be a lot.

Some technical nuggets on getting it to "function" in React AND w server-side rendering (functino doesn't mean this is a good solution going forward):
- The only export runs code that loads polyfills and attaches event handlers to the DOM
  - This is great for the plain html/js use case when the user is just including the script on the page, but not good for integration with libraries that manipulate the DOM like React or server rendering
  - Solution: export the components and polyfills so the implementor has control over DOM attachment/removal and loading of polyfills
  - Hack: see [webpack config](https://github.com/18F/federalist-gatsby-uswds-template/blob/master/gatsby-node.js#L13) and [webpack config](https://github.com/18F/federalist-gatsby-uswds-template/blob/master/gatsby-node.js#L47)
- There are polyfills included by dependencies that expect the DOM to be available when imported/required
  - Same as above, but in this case we can't directly separate them out
  - Solution: change the dependencies or conditionally require them only in the browser
  - Hack: see [webpack config](https://github.com/18F/federalist-gatsby-uswds-template/blob/master/gatsby-node.js#L31) and [shim](https://github.com/18F/federalist-gatsby-uswds-template/blob/master/shims/element-closest.js)

**Comment by thisisdano** [comment]
To me, this is around understanding "modern component architecture" from at least three perspectives:
- markup
- styling
- javascript

Perhaps you might also add 'data' to this list. Maybe other things?

I feel most comfortable understanding the markup and styling perspectives, and generally how props and data might flow and pass into components. But I don't really have a good technical handle on the js side, and this keeps me from really understanding how they all work together and the current pain points.

In general, I think that abstraction, modularity, and isolation are concepts we want to encourage in system architecture, in the interest of upgrade path, resilience, and compositional flexibility. Work that we do in the service of these goals has the possibility to be very valuable to the long-term health of the project

**Comment by duluca** [comment]
Had to create a sibling issue for Angular #2890. I think one thing that may be very helpful would be an update to README of this project, **allowing** React and Angular projects to use more suitable component libraries with native platform support that meet or exceed the guidelines. I did bold **allow**, because this repository is practically a legal document for contractors who are asked to adhere to USDWS guidelines/implementation. It'd go a long way to add some caveats in here.

**Comment by brockfanning** [comment]
I'm not a front-end developer so probably need to be educated here: Why do React (or Angular or whatever) components need to include the styles and javascript? Is anything lost by limiting the React/Angular/etc components to markup only, and including styles/javascript in the usual way, with link/script tags? Wouldn't that make the maintenance bit easier?

**Comment by duluca** [comment]
Modern frameworks allow for data-binding, which, in a sense, automatically connects the markup (view) to code behind the markup (controller or viewModel). As such the markup is written with specific syntax to make this happen. And taking this idea further, modern frameworks make it possible to combine both logic and markup in to reusable components.
On the upside, you are able to encapsulate the behavior, look and feel of the user control neatly in to reusable components, which makes it super easy to implement a design standard like USWDS. 
On the downside, each framework has a different way of implementing these components.
Looking in to the future, the web framework standard will standardize how components are actually consumed, so a component written in one framework can be used in others, but that future is not here today, which is why having USWDS components natively in React, Angular and other popular frameworks has an immense productivity gain for development teams, while ensuring the best possible and most consistent user experience for users of US government websites. 
My specific gripe out the current implementation is that it is locked in to how we thought about web design 5 years ago. If we pared it down to more simple HTML, CSS and JS that can be easily incorporated with modern frameworks, I think that's a viable direction to go as well.

**Comment by brockfanning** [comment]
Think I got it - so the javascript functionality might need to be included in the React/Angular/etc component to support reacting to changes in the model. So in the case of USWDS an example might be if you want to programmatically trigger the opening/closing of an "accordion" based on whatever state the model is in?

**Comment by tbaxter-18f** [comment]
I think I'm fairly closely aligned with @davemcorwin on this. I believe that 
* USWDS should provide simple vanilla JS where it's required for functionality
* USWDS should provide straightforward mechanisms to attach or unattach that vanilla JS. I was thinking a body class, but Dave has better ideas.
* USWDS should document those cases where JS is in place for transparency and to avoid conflicts

At that point, it would be relatively easy to stand up a USWDS-React, USWDS-Angular, USWDS-Vue, or USWDS-whatever project that detaches the default vanilla JS and instead leverages the desired framework. There may be some discussion to be had about additional hooks needed, but with good semantic attribute-rich markup (which the USWDS already has) there should be quite a few hooks in place already.

**Comment by darthrellimnad** [comment]
Love this conversation! ❤️ I'm pretty familiar with modern React, although my CSS skills are a bit dated now, but familiar enough with some of the techniques used in various communities.  I agree with comments here that biggest issue would be maintenance and matching updates with the vanilla version for updates :thinking:.

I'm wondering what you all think about something like [react-styleguidist](https://github.com/styleguidist/react-styleguidist) or similar tools. Could be an idea for incrementally implementing the styleguide as matching React components (or similar for Angular or Vue).  Probably easiest to start with the basics to start with (buttons, headers, layouts, etc). Converting the HTML to JSX isn't the hard part, but finding an automated way to update styles from vanilla source might be difficult, especially for more complex stateful UI and CSS like the accordian example... I'd have to think on it.  Might be a place to start though 🤔

**Comment by darthrellimnad** [comment]
Another (maybe short term) option could be to just use the produced css directly with decoupled React components?  Depending on the level of customization the Components need to be provided via [props](https://reactjs.org/docs/components-and-props.html#props-are-read-only), the react components could bake in the classNames, so consumers would just need to ensure to include the css file normally, and then import and use the Components, referencing public api & styleguide as necessary.

Problem with this approach would be optimizing for bundle sizes.  A lot of React developers us the [create-react-app](https://facebook.github.io/create-react-app/) tool to bootstrap an SPA or PWA project, which uses [Webpack](https://webpack.js.org/) internally.  Webpack can [tree shake](https://webpack.js.org/guides/tree-shaking/) unused code with static analyzers, so lots of people use this feature to remove unused code in production builds of their applications.

If the CSS is global and all loaded by the browser up-front, then we can't really optimize bundle sizes for the style or imported assets from the stylesheet.  To resolve, i think it would require co-locating the React Component JSX (or source templates/html) with a [css-module](https://github.com/css-modules/css-modules) or using one of the many [css-in-js](https://reactjs.org/docs/faq-styling.html) solutions.  These sorts of solutions effectively automate the [BEM](http://getbem.com/introduction/) pattern used to avoid name collisions in stylesheets.

While Webpack may be useful for consumers, (and likely a local dev/integration/test environments for style-guide & standards designers/developers) it may be build the open source npm distribution for the React components using something like [rollup](https://github.com/rollup/rollup).

**Comment by dwagner4** [comment]
I make two suggestions

1.  Don't forget Vue.js  up and coming framework
2.  Maybe looking into Google's lit-html element which can be extended to make framework agnostic light weight components is worth the effort.  https://lit-element.polymer-project.org/guide  You can learn it in an afternoon

**Comment by stphnwlkr** [comment]
I will add my opinion. I do not think it is the responsibility of the USWDS to provide anything more than the core layout, visual styling, JavaScript functionality, and best practices guides. If the user community wants to extend it for the framework of the month then what might be nice is an official  location for sharing extended features, component, framework implementations, themes, etc.

**Comment by tbaxter-18f** [comment]
I think it would be helpful if the parent issue was reframed a bit. I think there's pretty clear consensus here that the USWDS itself should be framework-agnostic, so it should NOT provide React components itself. I think the actual desire is to have the USWDS play nicely with frameworks in general, to not preclude or complicate their use, and to provide a good foundation for framework-specific community efforts to build upon as desired.

To do so, USWDS will have to provide generalized vanilla JS for "standard" usage, along with some mechanism to NOT load that JS if another framework is in use. it would then be up to that implementor or the community to provide framework-specific replacement JS.

I would suggest rewriting the issue and title to better reflect that desire. As is, it reads as though USWDS has an intent to be a react-specific system, which I don't believe aligns well with the project goals overall.

**Comment by darthrellimnad** [comment]
I agree with comments here about uswds responsibilities regarding frameworks. I'm mainly considering the maintenance cost for matching updates with vanilla source for each framework that has a community dedicated to its development.

I think most of the popular component frameworks these days could benefit from a great library of accessible web components, like those developed by uswds.  Specifically, the attention to detail and automated testing scripts are things we would (ideally) like for the framework specific versions as well.

So while I don't believe it's uswds responsibility to maintain framework specific libraries, there might be ways to organize the code and/or styles to help framework community developers when applying updates to match changes made to vanilla source. I'm mostly familiar with React currently, but have used Angular and Vue a bit, and I think some of these issues may be similar across all frameworks interested in creating a component library from uswds source. Worth discussion, probably with input from people more familiar with other frameworks too :)

**Comment by thisisdano** [comment]
So as far as I can tell there are a couple main themes here:

1. USWDS should focus on responsive component styling, basic JS, and accessibility — as well as best practice guidance around design.
2. USWDS JS should play well with modern frameworks like React and Vue
3. We should organize our project files in a way that makes it easy for downstream developers to build with them in their desired framework — and encourage them to do this.

I do still consider the points I raised in the initial issue to be significant:

> 1. It makes it very difficult to upgrade components and keep them in sync with the most current version of the design system

It is hard to track component changes from version to version and keeping markup/styles in sync with the current version of USWDS is a real challenge. One thing we can do in the short term here is try to figure out how we can do a better job of documenting these changes on a component-by-component basis — and perhaps encourage development of extensible templates written with methodologies that allow the end user to stay in sync with less irritating effort. (See the next point, below.)

> 2. It offers no abstraction layer between the component and the rendered html — leaving the component rather brittle and difficult to maintain.

I know that going from 1.x -> 2.0 is a pain in the neck. Going through and find/replacing class names is no way to run a website. Folks who build sites could and should do it in such a way that "you get the markup for free" — that is, implemented with some layer of abstraction where you're rarely dealing with adding classes as much as adding _components_ and passing values into them. While we may not provide React, etc components, thinking about components as markup + data/props (and perhaps documenting them more like this) will be useful to us.

> 3. It offers no explicit, practical guidance around best-practice implementation of USWDS components.

We will have to work with the developer community to improve our guidance around implementation. There's a lot to parse just in this discussion. I would love to work together to get to a point where we can support things like React and Vue implementations and keep them fresh, extensible, and useful to the wider gov community (including USWDS itself — I see no reason why the USWDS site couldn't use something like a React implementation alongside Gatsby and Federalist for our own site).

**Comment by tbaxter-18f** [comment]
For what it's worth, while I believe USWDS should have no opinion on frameworks, I believe it should be highly opinionated on markup. There just aren't that many ways to mark something up correctly once standards, semantics, a11y and other concerns are brought into mind. And I think highly opinionated HTML is highly predictable HTML, and is easier to port into other frameworks and have it work correctly.

**Comment by thisisdano** [comment]
Fair point @tbaxter-18f — I think we're doing an ok job with that now, and will try to do even better in the future. Good html markup is the foundation, overlaid with a consistent component class naming system. 

FWIW, starting with USWDS 2.0 we will be using BEM consistently throughout our class names. I know this doesn't explicitly address semantics, accessibility, etc — but it will impose some rigor and predictability on our component classing.

While USWDS often seems to focus on the presentation layer, I do believe that accessible markup is  true core that makes presentational accessibility (usability, readability, beauty, etc) possible.

**Comment by frankstallone** [comment]
Great conversation! I wanted to add my 2 cents as succinct as possible. I fully support building out the USWDS into relative popular frameworks as in my experience it has proven to increase adoption. It lowers the boundaries of entry into a fantastic well thought out design system. Take these suggestions with a grain of salt. I understand the implications for maintenance and think y'all are doing a great job already.

We have a Visual Design Language at ADP where there is a base CSS file, vanilla JS solutions to common UI components and it is also built out in React and Angular. Personally I'd love to re-design a few of my client's sites using USWDS 2.x in [WordPress](https://wordpress.org/) or React (via [Gatsby](https://www.gatsbyjs.org/). 

Via @darthrellimnad 
> I'm wondering what you all think about something like [react-styleguidist](https://github.com/styleguidist/react-styleguidist) or similar tools. 

I am partial to [Storybook](http://storybook.js.org/) myself. From my experience it's a nice agnostic tool for isolated components, similar to `react-styleguideist` but not tied to any one framework. 

> If the CSS is global and all loaded by the browser up-front, then we can't really optimize bundle sizes for the style or imported assets from the stylesheet.

I second your solution posed above by Daniel and I would consider [emotion](http://emotion.sh/) to encapsulate your CSS into the component. Don't use the component? Tree shaken out of the build, totally! Something I intend to champion internally @ ADP.

> Not sure if an automated (or partially automated) way to convert source to framework specific templating/component solutions would be practical though 🤔

This reminded me, Salesforce created [Theo](https://github.com/salesforce-ux/theo) to automate their design tokens. 

On the idea of CSS abstraction so one isn't updating from 1.x to 2.x and doing a massive find/replace I am intrigued by [Tailwind CSS](https://tailwindcss.com/docs/what-is-tailwind/)'s abstraction methodology. Not sure if it is relevant for a project like this but it's the best agnostic abstraction I have seen for CSS in some time. I am a fond BEM user.

**Comment by maxxheth** [comment]
Perhaps creating and standardizing USWDS web components via something like [StencilJS](https://stenciljs.com/) that can be dropped in a plug-and-play fashion regardless of one's choice of framework (or no framework at all) would be an approach worth exploring.

**Comment by duluca** [comment]
I'd like to second the suggestion from @maxxheth. As @tbaxter-18f puts it, it is really important to be opinionated (without being framework specific) however markup alone only captures part of the story that related to representation. HTML & CSS alone can't capture the full UX aspect & intended behavior.  Custom Elements (aka Web Components) can do this! Reusable React, Angular and Vue component does do this. 

In addition, HTML & CSS alone can be opinionated in all the wrong ways that makes implementing them costly and unfeasible in React, Angular and Vue, which are all opinionated in their own ways. 

I'd really love to see Version 3 of USWDS to be completely focused on reusable,drop-in Custom Elements that projects can just use with their framework of choice and don't have to worry about implementing 508 compliance from scratch and style guide is baked in to these components. IMO, this is the only way to achieve wide adoption of USWDS (which could even extend beyond the Federal Government).

**Comment by thisisdano** [comment]
@duluca your suggestion is very much on our radar

**Comment by duluca** [comment]
@mhartington can bring a lot of perspective to this discussion, working deep in reusable components for years and very closely with Stencil and Ionic.

**Comment by mhartington** [comment]
👋 Hi there! Thanks for tagging me @duluca

I'm from the Stencil team at Ionic. Looking over this issue, it sounds like Web Components (Custom elements) are probably the path you want to take. There's been a lot of good points made for Web Components, and if you're interested in stencil, I'm more than happy to talk more about it to the major players here.  This is the very reason we made stencil at Ionic, to bring our component library/design system to multiple frameworks.

**Comment by darthrellimnad** [comment]
@frankstallone

I like the look of tailwind too :) I've used tachyons before and really liked it for custom design prototyping, and tailwind's design seems a nice layer above that for extracting to classes, css-modules, or CSS in JS solutions.

Also, been playing with storybook some more since the update and really liking it too! 👍

@mharington

Really like the look of stencil, going to play with it a bit more when I have time.  You all did some great work on this! Thanks for sharing ❤

**Comment by loganpowell** [comment]
I would be interested in contributing to this effort

**Comment by darthrellimnad** [comment]
I finally got around to working with stencil a bit and I'm liking it so far!  Working with real custom elements and dropping them into a React app felt good.

A couple notes from my experience:
- writing stencil components felt awkward at first, especially since I was primarily writing JSX. I would forget that I wasn't in "React-land" and had to catch myself from time to time.
- when exporting reusable components for use in a separate project, I couldn't tell if there was a "recommended" way to render them on the server yet. I see some people recommending headless chrome (via [puppeteer](https://developers.google.com/web/tools/puppeteer/)), and a few other solutions, but seems like all custom elements may present some new challenges for SSR.  For now, my method is to prerender thin page wrapper(s) and allow content to be rendered upon page load in browser.
- I missed having React hooks available to me when building stateful stencil components... not a problem w/ stencil, more an issue of personal preference. I ended up writing mostly "stateless" functional stencil components, and wired them up with React hooks to supply props.  Side note, this led me to discover [haunted](https://github.com/matthewp/haunted) which looks promising!
- the context switchimg between react and stencil was a little tough for me.  I think if I was working in a self contained project with little chance of needing to reuse the components elsewhere, it may be simpler to write project-reusable components using the same framework.  For design systems however, stencil feels powerful!

All in all, seems like a great tool if the goal is to support popular JS frameworks.  If the goal was simply to support custom-elements and let framework developers build their own wrapping components, I'd also consider using [svelte](https://svelte.dev/).

**Comment by afeld** [comment]
Surprised this wasn't posted already: https://github.com/n8mellis/uswds-react

**Comment by ClydeDroid** [comment]
This one looks really nice: https://github.com/paularce/react-uswds2

**Comment by afeld** [comment]
https://github.com/trussworks/react-uswds

**Comment by aduth** [comment]
We use USWDS in our React project, and while a first-party React component library would certainly be convenient, I agree with the prior comments about the extra maintenance burden and lack of portability between frameworks.

Custom elements could well be the future, but aside from some [legitimate concerns](https://dev.to/richharris/why-i-don-t-use-web-components-2cia), I imagine it'd still be a big shift for the project to add and maintain these new component implementations.

In the meantime, the biggest struggle I've had with using USWDS in React is that many components don't work when elements are added after the initial page load. Some behaviors _do_ work thanks to event delegation via [`receptor`](https://www.npmjs.com/package/receptor), but some components must be initialized, and this initialization isn't (or can't be) run for newly added elements.

I've been stewing on this problem for a while, and I might wonder if [`MutationObserver`](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) could serve as a near-term stand-in for something akin to custom elements, to at least manage when component elements are added or removed from the page.

In quick-and-rough prototyping, I've had some success with such an approach:

- https://codepen.io/aduth/pen/bGWNJgZ
- https://codepen.io/aduth/pen/gOWbNPg

I since discovered there's even a few libraries which already exist based on this idea:

- https://github.com/WebReflection/regular-elements
- https://github.com/WebReflection/wicked-elements

(via blog post: https://webreflection.medium.com/a-wicked-custom-elements-alternative-6d1504b5857f)

This has a few advantages:

- Entirely backward-compatible, building on existing component markup.
- Good _future_ compatibility, since the public interface is limited to the standard (BEM) component class name, and the internal implementation can remain private and receive breaking changes without much downstream concern.
- [Browser support](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver#browser_compatibility) back to Internet Explorer 11 without polyfills.
- No lock-in, neither to a specific JavaScript framework, nor to a future transition toward custom elements. It may even make such a transition easier, if the component implements methods with a similar interface to custom elements (`connectedCallback`, etc).

And some drawbacks:

- On its own, `MutationObserver` doesn't allow for much precision to the concerned elements, so performance of the observer callback should be thoughtfully considered.
- If it's only a temporary solution, where custom elements are the ultimate goal, it may not be worth the invested effort.
- Framework consumers still need to manage the component HTML, which is more verbose and error-prone than a mostly self-contained custom element e.g. `<usa-date-picker />`.

**Comment by davemcorwin** [comment]
@aduth Does this approach assume that your application is using the USWDS javascript alongside your React code?

I think what would be helpful here is a cleaner separation of the javascript from the rest of the library along with documentation about what needs to be done via javascript. Then users can chose the approach that best fits the architecture of their application.

**Comment by aduth** [comment]
@davemcorwin Yeah, the assumption in that example would be that USWDS JavaScript would need to be loaded to add behaviors for component elements on the page (including after initial load), and that the consumer would be responsible for getting that markup into the page in the first place (via React, server-side rendering, etc.).

Could you elaborate more on what you mean by cleaner separation? I'm not sure I follow what the concern or use-case is.

**Comment by davemcorwin** [comment]
The architecture of USWDS was designed for traditional server-rendered or pre-rendered (Jekyll, Hugo, etc) applications and not necessarily with a componentized approach in mind. If we just focus on the javascript, this is why it includes polyfills, code with global effects (not just components), and a seldom-used [event delegation library](https://github.com/shawnbot/receptor).

When using the USWDS JS  alongside other JS frameworks and/or in single-page applications, this can lead to code duplication, conflicts, or broken functionality.

At some point USWDS should decide whether or not the USWDS js is *intended* to be used:
- in entirety alongside other js
- in entirety only by itself
- in entirety only by itself OR in components alongside other js

Your suggestion is relevant to the first option, but even if the code were refactored to use something like `MutationObserver` there would continue to be edge cases relevant to whatever JS flavor of the week is being used alongside of it that USWDS just doesn't have to resources to address.

Alternatively, USWDS could explicitly say that the JS should only be used by itself, but could provide documentation on what needs to be done for the components to work and maintain 508 compliance.

Lastly, USWDS could refactor the code and provide additional documentation to allow the individual component js to be used alongside another js library. I have examples of doing this with React and it sorta worked, but I don't think this is the best approach.

I *think* the second option would be my favorite approach, but USWDS could look at reimplementing the JS in something like [Alpine.js](https://alpinejs.dev/) which would make it very clear in the examples what the JS needs to do.

I thought the Tailwind folks had done some work like this, but now when I look at [TailwindUI](https://tailwindui.com/) I only see React and Vue via [HeadlessUI](https://headlessui.dev/) and no more Alpine.js.

**Comment by aduth** [comment]
> * in entirety alongside other js
> * in entirety only by itself
> * in entirety only by itself OR in components alongside other js
> 
> Your suggestion is relevant to the first option

I'm not suggesting that USWDS _must_ be paired with a developer's own JavaScript. My main concern is that it should support that it _could_ be used in a JavaScript application if one chooses. Today, many components simply do not work when used in React or any other JavaScript framework.

**Comment by mejiaj** [comment]
Love where this discussion is going. I like this solution in the short term:
> Alternatively, USWDS could explicitly say that the JS should only be used by itself, but could provide documentation on what needs to be done for the components to work and maintain 508 compliance.

Alpine seems interesting, but the markup is already problematic.
>I think the second option would be my favorite approach, but USWDS could look at reimplementing the JS in something like Alpine.js which would make it very clear in the examples what the JS needs to do.

One of the reasons @thisisdano outlined earlier is that it's hard to track component changes in markup and keep those synced. 

Web components make the most sense to me based on the problems mentioned. Maybe we can serve the compiled markup as an example for those on traditional sites.

**Comment by davemcorwin** [comment]
@aduth You can take a gander at https://github.com/18F/federalist-uswds-gatsby/tree/v1 for an example of trying to just integrate the USWDS component js in a React app. I had to do some Webpack tomfoolery and other stuff. Note that this tag is no longer used or maintained.

I didn't think you were saying the first part of this
> I'm not suggesting that USWDS must be paired with a developer's own JavaScript.

I was just trying to address
> My main concern is that it should support that it could be used in a JavaScript application if one chooses. Today, many components simply do not work when used in React or any other JavaScript framework.

The USWDS team has to decide which route they want to take and a huge driver is going to be the amount of work to get there and maintain it. The work required to cleanly separate the component code from the global stuff, allow easy use of either way, and maintain compatibility with every JS library will probably outweigh that of keeping the status quo while adding documentation and guidance. I think this fits well with what is currently done for the rest of the library.

**Comment by davemcorwin** [comment]
We could also look at leveraging or taking inspiration from something like https://reach.tech/ which provides 508 compliant React components which maybe could be styled with USWDS?

**Comment by jkjustjoshing** [comment]
Not sure the best place to post an example of a design system built with web components - here's one that might be good inspiration https://auro.alaskaair.com
